spec/ics-3-connection-semantics/README.md

```diff
@@ -0,0 +1,395 @@
+
+`Identifier` and other host state machine requirements are as defined in [ICS 24](../spec/ics-24-host-requirements). The identifier is not necessarily intended to be a human-readable name (and likely should not be, to discourage squatting or racing for identifiers).
+
+The opening handshake protocol allows each chain to verify the identifier used to reference the connection on the other chain, enabling modules on each chain to reason about the reference on the other chain.
+
+A *actor*, as referred to in this specification, is an entity capable of executing datagrams who is paying for computation / storage (via gas or a similar mechanism) but is otherwise untrusted. Possible actors include:
+- End users signing with an account key
+- On-chain smart contracts acting autonomously or in response to another transaction
+- On-chain modules acting in response to another transaction or in a scheduled manner
+
+### Desired Properties
+
+- Implementing blockchains should be able to safely allow untrusted actors to open and update connections.
+
+#### Pre-Establishment
+
+Prior to connection establishment:
+
+- No further IBC subprotocols should operate, since cross-chain substates cannot be verified.
+- The initiating actor (who creates the connection) must be able to specify an initial consensus state for the chain to connect to and an initial consensus state for the connecting chain (implicitly, e.g. by sending the transaction).
+
+#### During Handshake
+
+Once a negotiation handshake has begun:
+
+- Only the appropriate handshake datagrams can be executed in order.
+- No third chain can masquerade as one of the two handshaking chains
+
+#### Post-Establishment
+
+Once a negotiation handshake has completed:
+
+- The created connection objects on both chains contain the consensus states specified by the initiating actor.
+- No other connection objects can be maliciously created on other chains by replaying datagrams.
+- The connection should be able to be voluntarily & cleanly closed by either blockchain.
+- The connection should be able to be immediately closed upon discovery of a consensus equivocation.
+
+## Technical Specification
+
+### Data Structures
+
+This ICS defines the `Connection` type:
+
+```golang
+type ConnectionState enum {
+  INIT
+  TRYOPEN
+  TRYCLOSE
+  OPEN
+  CLOSED
+}
+```
+
+```golang
+type Connection struct {
+  state                         ConnectionState
+  counterpartyIdentifier        Identifier
+  clientIdentifier              Identifier
+  counterpartyClientIdentifier  Identifier
+  nextTimeoutHeight             uint64
+}
+```
+
+### Subprotocols
+
+This ICS defines three subprotocols: opening handshake, header tracking, closing handshake. Datagrams defined herein are handled as external messages by the IBC relayer module defined in [ICS 26](../spec/ics-26-relayer-module).
+
+![State Machine Diagram](state.png)
+
```

Comment:
Nice diagram :) Could you make the text on transitions a bit bigger? It doesn't read easy currently.

Reply:
Also, IMHO it would be clearer if you'd separate A and B FSMs.

Reply:
Thanks for the feedback, makes sense - will do.

Reply:
Altered the font sizing. I'm not sure separating the FSMs for A and B is the right direction, though - the goal of the protocol construction is to allow reasoning about the combined state of the IBC handlers on both chains at once, so it seems like we want to combine them...

Reply:
...I agree we want to look at chains' FSMs at once. But the current diagram seem to show some transitions upon events that are not valid one one side. Let me try to read it more carefully the way it is together with the code.

Reply:
I think we should list the events in the FSM (CONNINIT, CONNOPENTRY, etc) and describe what triggers them. Maybe add also a column in the table below?

Reply:
Added two columns in each table, for prior state and posterior state - is that what you meant?